sw/device/lib/testing/usb_testutils_diags.h

```diff
@@ -0,0 +1,51 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_TESTING_USB_TESTUTILS_DIAGS_H_
+#define OPENTITAN_SW_DEVICE_LIB_TESTING_USB_TESTUTILS_DIAGS_H_
+// Diagnostic, testing and performance measurements utilities for verification
+// of usbdev and development of the usb_testutils support software; the
+// requirements of this software are peculiar in that the USBDPI model used in
+// top-level requires packet responses very promptly, so the introduction of
+// logging/tracing code can substantially alter behavior and cause malfunction
+//
+// Employ faster memory copying routines to/from the packet buffer
+//   (the standard dif_usbdev_buffer_write/read functionality should normally be
+//    employed, but it is inefficient; the replacement routines better model the
+//    attainable performance)
+#define USBUTILS_MEM_FASTER 0
+
+// Implement low-impact tracing of software execution, permitting the software
+//   and hardware behavior to be married, hopefully without modifying the
+//   behavior/performance, particularly in top-level simulation
+#define USBUTILS_FUNCTION_POINTS 0
+
+// Record the function points to a memory buffer instead, for use where test
+//   hardware is unavailable, eg. FPGA builds
+#define USBUTILS_FUNCPT_USE_BUFFER 0
+
+#if USBUTILS_FUNCTION_POINTS
+// Omitted from source
+#else
+// Omit function point tracing
+#define USBUTILS_FUNCPT(pt, d)
+#endif
```

Comment:
Addressed. I had not fully separated out the diags (one PR, not yet created) and streaming (another, extant PR) when trying to simplify things.
I definitely want to ensure that the diags code is included at some point, however, since it is excruciating working at the level of t-l sim. The build and simulation times are so long, and visibility into the activity and timing of the software is extremely poor without the diags I have found it necessary to create.
Alas the timings of the sw and the DPI model are still closely interlinked, just less so than earlier.

Reply:
Providing no 'test descriptor' (NULL desc pointer) means that the usb_testutils layer will not respond to the command from the DPI model. That's fine for the hello_usbdev example application at present, since it only runs on the FPGA, but the top-level simulation tests must provide a response or the DPI model will halt; this is more robust than having it make assumptions in the event of invalid or absent response.

Reply:
This makes sense, thanks for the clarification @alees24 .